torch/_inductor/fx_passes/quantization.py

```diff
@@ -903,38 +929,40 @@ def clone_to_new_node(graph, source_node, user_node):
                 user_node.replace_input_with(source_node, new_node)
             return new_node
 
-        if dtype == torch.float32:
-            mul_node = match.output_node()
-        else:
-            convert_to_bf16_node = match.output_node()
-            mul_node = convert_to_bf16_node.args[0]
-        sub_node = mul_node.args[0]
-        to_fp32_node = sub_node.args[0]
-        assert mul_node.target is aten.mul.Tensor
-        assert sub_node.target is aten.sub.Tensor
-        assert to_fp32_node.target is prims.convert_element_type.default
+        # Find the start node and end node of a dequant pattern to be cloned
+        # * End node should be the match.output_node()
+        # * Start node should be the dtype convert to float32 node
+        dequant_pattern_end_node = match.output_node()
+        assert dequant_pattern_end_node.target in [
+            aten.mul.Tensor,
+            prims.convert_element_type.default,
+            aten.reshape.default,
+        ]
 
-        graph = match.graph
-        user_node_list = (
-            list(mul_node.users)
-            if dtype == torch.float32
-            else list(convert_to_bf16_node.users)
-        )
-        for user_node in user_node_list:
-            # Step1: Duplicate the mul node
-            if dtype == torch.float32:
-                new_mul_node = clone_to_new_node(graph, mul_node, user_node)
+        def _find_first_node_in_dequant_pattern(_node):
+            if (
+                _node.target is prims.convert_element_type.default
+                and _node.args[1] == torch.float32
+            ):
+                # For a dequant pattern, we expect the start node is a to_fp32 node
+                return _node
             else:
-                new_convert_to_bf16_node_node = clone_to_new_node(
-                    graph, convert_to_bf16_node, user_node
-                )
-                new_mul_node = clone_to_new_node(
-                    graph, mul_node, new_convert_to_bf16_node_node
-                )
-            # Step2: Duplicate the sub node
-            new_sub_node = clone_to_new_node(graph, sub_node, new_mul_node)
-            # Step3: Duplicate the to_fp32 node
-            _ = clone_to_new_node(graph, to_fp32_node, new_sub_node)
+                return _find_first_node_in_dequant_pattern(_node.args[0])
```

Comment:
do we need to assert that args[0] always valid?

Reply:
Thanks. Add the assert of args' length here.